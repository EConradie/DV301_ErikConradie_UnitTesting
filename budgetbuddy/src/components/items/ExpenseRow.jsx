import React from 'react'

function ExpenseRow(props) {

  const {expense} = props;

  return (
    <div className='card card-row p-3 mb-2 row-expense' data-testid="expense-card">
        <p aria-label='title' >{expense.title}</p>
        <p aria-label='amount' className='amount'>R {expense.amount.toFixed(2)}</p>
    </div>
  )
}

export default ExpenseRow