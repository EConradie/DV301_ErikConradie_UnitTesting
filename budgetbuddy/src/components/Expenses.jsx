import React, { useState } from "react";
import { Form, Button } from "react-bootstrap";
import ExpenseRow from "./items/ExpenseRow";

function Expenses({ expenses: initialExpenses, handleNewExpense }) {
  const [expenses, setExpenses] = useState([...initialExpenses]);
  // Adding state for each form field
  const [title, setTitle] = useState("");
  const [amount, setAmount] = useState("");

  const handleAddExpense = (event) => {
    event.preventDefault();
    const amountNumber = Number(amount); // Convert amount to number
    if (title && amountNumber) { // Check if title and amount are provided
      const newExpense = { title, amount: amountNumber };
      setExpenses([...expenses, newExpense]);
      handleNewExpense(newExpense);
      // Reset form fields using state
      setTitle("");
      setAmount("");
    }
  };

  return (
    <div data-testid="expenses-component">
      <h3>Household Monthly Expenses</h3>
      <Form onSubmit={handleAddExpense}>
        <div className="form-row">
          <Form.Control
            type="text"
            aria-label="title"
            name="title"
            placeholder="Expense Title"
            autoComplete="off"
            value={title} // Controlled component
            onChange={(e) => setTitle(e.target.value)} // Update state on change
          />
          <Form.Control
            type="number"
            aria-label="amount"
            name="expense"
            step="0.01"
            placeholder="0.00"
            autoComplete="off"
            value={amount} // Controlled component
            onChange={(e) => setAmount(e.target.value)} // Update state on change
          />
          <Button aria-label="button" type="submit" className="add-expenses">
            Add
          </Button>
        </div>
      </Form>

      <div className="expense-outer hide-scroll">
        {expenses.map((item, index) => (
          <ExpenseRow key={index} expense={item} />
        ))}
      </div>
    </div>
  );
}

export default Expenses;
