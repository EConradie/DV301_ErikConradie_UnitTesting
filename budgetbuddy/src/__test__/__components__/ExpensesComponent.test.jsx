import { render, screen } from "@testing-library/react";
import Expenses from "../../components/Expenses";
import ExpenseRow from "../../components/items/ExpenseRow";
import userEvent from "@testing-library/user-event";

test("Test default state of Expenses Component", () => {
  render(<Expenses expenses={[]} />);

  var fullComponent = screen.getByTestId("expenses-component");
  var title = screen.getByText("Household Monthly Expenses");
  var amount = screen.getByLabelText("amount");
  var button = screen.getByRole("button");

  expect(title).toBeInTheDocument();
  expect(amount).toBeInTheDocument();
  expect(button).toBeInTheDocument();

  expect(fullComponent).toMatchSnapshot();
});

test("Test if expense cards display when 2 expenses are added", () => {
  var expenses = [
    {
      title: "Expense 1",
      amount: 1000,
    },
    {
      title: "Expense 2",
      amount: 1000,
    },
  ];

  render(
    <div>
      {expenses.map((expense, index) => (
        <ExpenseRow key={index} expense={expense} />
      ))}
    </div>
  );

  var titleDisplay = screen.getAllByLabelText("title");
  var amountDisplay = screen.getAllByLabelText("amount");

  var expenseCards = screen.getAllByTestId("expense-card");

  expenseCards.forEach((element, index) => {
    expect(element).toBeInTheDocument();
    expect(element).toHaveTextContent(`R ${expenses[index].amount.toFixed(2)}`);
    expect(element).toHaveTextContent(expenses[index].title);
  });

  expect(titleDisplay).toHaveLength(2);
  expect(amountDisplay).toHaveLength(2);
});

test("Test adding of a new expense", async () => {
  var user = userEvent.setup();

  var expenses = [];

  const handleNewExpense = (newExpense) => {
    expenses = [...expenses, newExpense];
  };

  render(<Expenses expenses={expenses} handleNewExpense={handleNewExpense} />);

  var titleInput = screen.getByLabelText("title");

  await user.click(titleInput);
  await user.keyboard("test");

  expect(titleInput).toHaveValue("test");

  var amountInput = screen.getByLabelText("amount");

  await user.click(amountInput);
  await user.keyboard("1000");

  expect(amountInput).toHaveValue(1000);

  var button = screen.getByRole("button");

  await user.click(button);

  expect(expenses.length).toBe(1);
  expect(expenses[0].title).toBe("test");
  expect(expenses[0].amount).toBe(1000);
});
