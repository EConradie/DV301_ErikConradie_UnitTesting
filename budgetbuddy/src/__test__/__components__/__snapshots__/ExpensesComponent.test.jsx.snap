// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test default state of Expenses Component 1`] = `
<div
  data-testid="expenses-component"
>
  <h3>
    Household Monthly Expenses
  </h3>
  <form
    class=""
  >
    <div
      class="form-row"
    >
      <input
        aria-label="title"
        autocomplete="off"
        class="form-control"
        name="title"
        placeholder="Expense Title"
        type="text"
        value=""
      />
      <input
        aria-label="amount"
        autocomplete="off"
        class="form-control"
        name="expense"
        placeholder="0.00"
        step="0.01"
        type="number"
        value=""
      />
      <button
        aria-label="button"
        class="add-expenses btn btn-primary"
        type="submit"
      >
        Add
      </button>
    </div>
  </form>
  <div
    class="expense-outer hide-scroll"
  />
</div>
`;
